Microsoft.Toolkit.Mvvm/ComponentModel/ViewModelBase.cs

```diff
@@ -0,0 +1,193 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#pragma warning disable SA1512
+
+// Original file header:
+// ****************************************************************************
+// <copyright file="ViewModelBase.cs" company="GalaSoft Laurent Bugnion">
+// Copyright © GalaSoft Laurent Bugnion 2009-2016
+// </copyright>
+// ****************************************************************************
+// <author>Laurent Bugnion</author>
+// <email>laurent@galasoft.ch</email>
+// <date>22.4.2009</date>
+// <project>GalaSoft.MvvmLight</project>
+// <web>http://www.mvvmlight.net</web>
+// <license>
+// See license.txt in this project or http://www.galasoft.ch/license_MIT.txt
+// </license>
+// <LastBaseLevel>BL0014</LastBaseLevel>
+// ****************************************************************************
+
+using System.Runtime.CompilerServices;
+using Microsoft.Toolkit.Mvvm.Messaging;
+using Microsoft.Toolkit.Mvvm.Messaging.Messages;
+using Microsoft.Toolkit.Mvvm.Services;
+
+namespace Microsoft.Toolkit.Mvvm.ComponentModel
```

Comment:
There seems to be a pretty deep structure to the namespaces, since someone may be using multiple parts of these in the same classes, will that be a bit excessive? Just wondering if it's worth flattening this out at all? Thoughts?

Reply:
That's a good point! I've made some tweaks in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3229/commits/0d80519dd1e8a4efaa0a96fae5f47e3e24de635f and https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3229/commits/900cc9884060593bb66170ced9bb6e9c30e78258, I'll go ahead and explain my rationale here as to why I structured the namespaces the way they are, and then you can let me know what you think!

Right now the PR is structured like so:

- `Microsoft.Toolkit.Mvvm.Input`: contains all the commands and command interfaces. This effectively mirrors the namespace of the `ICommand` interface itself, which lives in `System.Windows.Input`.
- `Microsoft.Toolkit.Mvvm.ComponentModel`: contains the observable base classes (`ObservableObject` and `ViewModelBase`). I picked this name because those types are heavily relying on the `INotifyPropertyChanged` interface, which itself lives in `System.ComponentModel`.
- `Microsoft.Toolkit.Mvvm.DependencyInjection`: contains the `Ioc` types. This mirrors the `Microsoft.Extensions.DependencyInjection` namespace, which contains the ASP.NET Ioc container.
- `Microsoft.Toolkit.Mvvm.Messaging`: contains the `Messenger` types. Placed this in a separate namespace since some users might not be interested in this type/pattern, so they could just not import those types and hide them away.

In general, except from the `Messenger` type which is exclusive to this lib (and `MvvmLight`), I essentially replicated the following translation, hoping it could be intuitive for devs too:
- `System.Windows.Input` ---> `Microsoft.Toolkit.Mvvm.Input`
- `System.ComponentModel` ---> `Microsoft.Toolkit.Mvvm.ComponentModel`
- `Microsoft.Extensions.DependencyInjection` ---> `Microsoft.Toolkit.Mvvm.DependencyInjection`

That said, any suggestions are welcome, if you feel like I should make any other changes just let me know and I'll be happy to help! 😊

Reply:
Thanks @Sergio0694 that's great rationale and makes sense now. Will be great for docs around this if we move forward. 😊

Reply:
The rationale makes sense, but if the intention is to have only a few classes, then let's keep one namespace. We shouldn't have namespaces with one class.

Reply:
For reference, this is the current structure of the package (note that not all subfolders also have a namespace, eg. the command interfaces are still in `*.Input`):

![image](https://user-images.githubusercontent.com/10199417/79614875-c5b2c800-8101-11ea-9c6e-a5f6823c2a85.png)

This is just my two cents, but I don't really see why having more namespaces (which in this case are just 4 anyway) is inherently bad, especially if they're following step-by-step the existing architecture of known types from the BCL or from other 1st party packages (ie. `Microsoft.Extensions.DependencyInjection`). Having them separately would also allow devs not using them not to have all of them show up at the same time in the IntelliSense suggestion box when typing something.